Fatras/include/ActsFatras/Kernel/InteractionList.hpp

```diff
@@ -0,0 +1,344 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2018-2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Material/MaterialSlab.hpp"
+#include "ActsFatras/EventData/Particle.hpp"
+
+#include <bitset>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+namespace ActsFatras {
+namespace detail {
+
+/// Retrieve index of the first matching type in the tuple.
+///
+/// Taken from https://stackoverflow.com/a/18063608.
+template <class T, class Tuple>
+struct TupleIndexOf;
+template <class T, class... Types>
+struct TupleIndexOf<T, std::tuple<T, Types...>> {
+  static constexpr std::size_t value = 0u;
+};
+template <class T, class U, class... Types>
+struct TupleIndexOf<T, std::tuple<U, Types...>> {
+  static constexpr std::size_t value =
+      1u + TupleIndexOf<T, std::tuple<Types...>>::value;
+};
+
+// Construct an index sequence for a subset of the tuple elements.
+//
+// Whether an element is part of the subset is defined by a configurable
+// selector template struct. It must take the element type as its only template
+// parameter and must provide a static `Value` member value. If the value
+// evaluates to `true`, then the corresponding index will be part of the index
+// sequence.
+//
+// Example: The tuple contains four elements, where all but the third one (i=2)
+// will be selected. This should lead to the following recursive expansion
+// where the index sequence of the subset is filled from the front.
+//
+//        TupleIndicesSubsetImpl<..., kCounter=4>          // select index=3
+//     -> TupleIndicesSubsetImpl<..., kCounter=3, 3>       // skip   index=2
+//     -> TupleIndicesSubsetImpl<..., kCounter=2, 3>       // select index=1
+//     -> TupleIndicesSubsetImpl<..., kCounter=1, 1, 3>    // select index=0
+//     -> TupleIndicesSubsetImpl<..., kCounter=0, 0, 1, 3> // terminate
+//
+template <template <typename> typename selector_t, typename tuple_t,
+          size_t kCounter, size_t... kIndices>
+struct TupleSubsetIndicesImpl {
```

Comment:
This is a bit of a nitpick, but maybe a simpler name for this would be `TupleFilterImpl`, and `predicate_t` instead of `selector_t`?

Reply:
Yes, that is simpler. I will change that.

Reply:
Renamed to `TupleFilterImpl` and `TupleFilter`.